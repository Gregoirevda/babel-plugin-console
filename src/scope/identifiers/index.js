const { negate } = require('../../utils');

const getIdentifiers = predicate => bindings =>
  Object.keys(bindings).map(key => bindings[key]).filter(predicate).map(({ identifier }) => identifier);

const isParam = binding => binding.kind === 'param';

const getParamIdentifiers = getIdentifiers(isParam);
const getVariableIdentifiers = getIdentifiers(negate(isParam));

export { getParamIdentifiers, getVariableIdentifiers };
