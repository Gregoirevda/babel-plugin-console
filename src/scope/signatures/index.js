const looksLike = require('../../utils/looks-like');

export default (types, parameters, rootFunction) => {
  if (parameters.length > 1) {
    return getCurriedFunctionSignature(types, parameters, rootFunction);
  }

  return getFunctionSignature(types, parameters, rootFunction);
};

const getCurriedFunctionSignature = (types, parameters, rootFunction) => {
  if (types.isFunctionDeclaration(rootFunction)) {
    const kind = 'const';
    const { column, line } = rootFunction.node.loc.start;
    const name = rootFunction.node.id.name;
    const params = buildCurriedParameters(parameters);

    return buildCurriedSignature(line, column, kind, name, params);
  }

  if (types.isExportDefaultDeclaration(rootFunction.parent)) {
    const { node: declaration } = rootFunction.findParent(par => types.isExportDefaultDeclaration(par));
    const { column, line } = declaration.loc.start;
    const params = buildCurriedParameters(parameters);

    return buildCurriedDefaultExportSignature(line, column, params);
  }

  const { node: declaration } = rootFunction.findParent(par => types.isVariableDeclaration(par));
  const kind = declaration.kind;
  const { column, line } = declaration.loc.start;
  const name = rootFunction.parent.id.name;
  const params = buildCurriedParameters(parameters);

  return buildCurriedSignature(line, column, kind, name, params);
};

const buildCurriedParameters = parameters => {
  return parameters.reduce((acc, arg, index) => {
    const stringArgs = arg.map(({ name }) => name).join(', ');
    if (index === 0) return `(${stringArgs})`;
    return `${acc} => (${stringArgs})`;
  }, '');
};

const buildCurriedSignature = (line, column, kind, name, params) =>
  `(${line}:${column}) ${kind} ${name} = ${params} => {...}`;

const buildCurriedDefaultExportSignature = (line, column, params) =>
  `(${line}:${column}) export default = ${params} => {...}`;

const getFunctionSignature = (types, parameters, rootFunction) => {
  const params = parameters[0].map(({ name }) => name).join(', ');

  if (looksLike(rootFunction, { node: { id: types.isIdentifier } })) {
    const name = rootFunction.node.id.name;
    const { column, line } = rootFunction.node.loc.start;
    return buildFunctionDeclarationSignature(line, column, name, params);
  }

  if (types.isExportDefaultDeclaration(rootFunction.parent)) {
    const { node: declaration } = rootFunction.findParent(par => types.isExportDefaultDeclaration(par));
    const { column, line } = declaration.loc.start;
    return buildDefaultExportSignature(line, column, params);
  }

  const name = rootFunction.parent.id.name;
  const { column, line } = rootFunction.parent.loc.start;
  const kind = rootFunction.findParent(par => types.isVariableDeclaration(par)).node.kind;

  if (types.isArrowFunctionExpression(rootFunction)) {
    return buildArrowFunctionSignature(line, column, kind, name, params);
  }

  return buildFunctionExpressionSignature(line, column, kind, name, params);
};

const buildFunctionDeclarationSignature = (line, column, name, params) =>
  `(${line}:${column}) function ${name}(${params}) {...}`;

const buildDefaultExportSignature = (line, column, params) =>
  `(${line}:${column}) export default function (${params}) {...}`;

const buildFunctionExpressionSignature = (line, column, kind, name, params) =>
  `(${line}:${column}) ${kind} ${name} = function(${params}) => {...}`;

const buildArrowFunctionSignature = (line, column, kind, name, params) =>
  `(${line}:${column}) ${kind} ${name} = (${params}) => {...}`;
