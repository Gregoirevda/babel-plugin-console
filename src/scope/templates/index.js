const buildGroupCollapsed = template => title => level => {
  const indentation = indentWith(level, '| ', 'â–¼ ');

  return template(`
    if (typeof window !== 'undefined') {
      console.groupCollapsed(title);
    } else {
      console.log('${indentation}', title);
    }
  `)({ title });
};

const buildGroupEnd = template => title =>
  template(`
  if (typeof window !== 'undefined') {
    console.groupEnd(title);
  }
`)({ title });

const buildLog = template => (...args) => level => {
  if (level == 0) {
    return template('console.log(args);')({ args });
  }

  const indentation = indentWith(level, '| ', '| ');
  return template(`
    if (typeof window !== 'undefined') {
      console.log(args);
    } else {
      console.log('${indentation}', args);

    }
  `)({ args });
};

const indentWith = (level, str, initial) => Array.from({ length: level }).reduce(acc => `${str}${acc}`, initial);

export default template => ({
  buildGroupCollapsed: buildGroupCollapsed(template),
  buildGroupEnd: buildGroupEnd(template),
  buildLog: buildLog(template)
});
