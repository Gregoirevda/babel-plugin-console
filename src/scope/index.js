import { flatten } from '../utils';
import looksLike from '../utils/looks-like';

import { getParamIdentifiers, getVariableIdentifiers } from './identifiers';
import getSignature from './signatures';
import templateBuilders from './templates';

export default (path, template, t) => {
  const { buildGroupCollapsed, buildGroupEnd, buildLog } = templateBuilders(template);

  const scope = traverseFunctions(path, t);

  if (!scope) {
    return;
  }

  const { parameterIdentifiers, parentScope, returnStatement, signature, variableIdentifiers } = scope;

  const mapIdentifer = identifier => {
    const { loc: { start: { column, line } }, name } = identifier;
    return buildLog(t.stringLiteral(`(${line}:${column})`), t.stringLiteral(`${name}:`), identifier);
  };
  const parameters = parameterIdentifiers.map(mapIdentifer);
  const variables = variableIdentifiers.map(mapIdentifer);
  const buildReturn = returnStatement => {
    if (returnStatement) {
      const { loc: { start: { column, line } } } = returnStatement;
      return buildLog(t.stringLiteral(`(${line}:${column})`), returnStatement);
    }
    return buildLog(t.stringLiteral('Void'));
  };
  const scriptScope = Object.keys(parentScope).map(key => {
    const { identifier } = parentScope[key];
    const { loc: { start: { column, line } }, name } = identifier;
    return buildLog(t.stringLiteral(`(${line}:${column})`), t.stringLiteral(`${name}:`), identifier);
  });

  return [
    buildLog(...path.node.arguments),
    buildGroupCollapsed(t.stringLiteral(signature)),
    buildGroupCollapsed(t.stringLiteral('Parameters')),
    ...parameters,
    buildGroupEnd(t.stringLiteral('Parameters')),
    buildGroupCollapsed(t.stringLiteral('Variables')),
    ...variables,
    buildGroupEnd(t.stringLiteral('Variables')),
    buildGroupCollapsed(t.stringLiteral('Return')),
    buildReturn(returnStatement),
    buildGroupEnd(t.stringLiteral('Return')),
    buildGroupCollapsed(t.stringLiteral('Script')),
    ...scriptScope,
    buildGroupEnd(t.stringLiteral('Script')),
    buildGroupEnd(t.stringLiteral(signature))
  ];
};

const traverseFunctions = (path, t) => {
  const parentFunction = findParentFunction(path, t);

  if (!parentFunction) {
    return null;
  }

  let rootFunction = parentFunction;
  let variables = [getVariableIdentifiers(parentFunction.scope.bindings)];
  let parameters = [getParamIdentifiers(parentFunction.scope.bindings)];

  while (isNotRootFunction(rootFunction, t)) {
    const maybeParentFunction = findParentFunction(rootFunction, t);
    if (!maybeParentFunction) {
      break;
    }
    rootFunction = maybeParentFunction;
    variables.unshift(getVariableIdentifiers(rootFunction.scope.bindings));
    parameters.unshift(getParamIdentifiers(rootFunction.scope.bindings));
  }

  const parentScope = rootFunction.scope.parent.bindings;
  const stmt = parentFunction.node.body.body.find(rest => t.isReturnStatement(rest));
  const returnStatement = stmt && stmt.argument;

  return {
    parameterIdentifiers: flatten(parameters),
    parentScope,
    returnStatement,
    signature: getSignature(t, parameters, rootFunction),
    variableIdentifiers: flatten(variables)
  };
};

const findParentFunction = (path, t) =>
  path.findParent(parent => {
    return t.isFunctionDeclaration(parent) || t.isArrowFunctionExpression(parent) || t.isFunctionExpression(parent);
  });

const isNotRootFunction = (path, t) =>
  !looksLike(path, { node: { id: t.isIdentifier } }) &&
  !t.isVariableDeclarator(path.parent) &&
  !t.isFunctionExpression(path.parent);
