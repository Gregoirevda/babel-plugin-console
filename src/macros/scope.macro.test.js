import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-macros';

pluginTester({
  plugin,
  snapshot: true,
  tests: withFilename({
    'Called outside of a function > remove console scope call': {
      snapshot: true,
      code: `
        import scope from './scope.macro'
        const c = 10;
        scope('should be removed');
      `
    },
    'Function Declaration > does not change when custom console scope is not present': {
      code: `
      import scope from './scope.macro'
      function add(a, b) {
        const c = 10;
        console.log('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Function Expression > does not change when custom console scope is not present': {
      code: `
      import scope from './scope.macro'
      const add = function (a, b) {
        const c = 10;
        console.log('Custom add function message');
        return a + b + c;
      };
      `
    },
    'Array Function > does not change when custom console scope is not present': {
      code: `
      import scope from './scope.macro'
      const add = (a, b) => {
        const c = 10;
        console.log('Custom add function message');
        return a + b + c;
      };
      `
    },
    'Curried Function > does not change when custom console scope is not present': {
      code: `
      import scope from './scope.macro'
      const add = a => b => {
        const c = 10;
        console.log('Custom add function message');
        return a + b + c;
      };
      `
    },
    'Function Declaration > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro'
      function add(a, b) {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Function Expression > const > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = function(a, b) {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Function Expression > let > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      let add = function(a, b) {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Arrow Function > const > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = (a, b) => {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Arrow Function > let > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      let add = (a, b) => {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Curried Arrow Function > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = (a) => (b) => {
        const c = 10;
        scope('Custom add function message');
        return a + b + c;
      }
      `
    },
    'Curried Function Declarion > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      function add(a) {
        return (b) => {
          const c = 10;
          scope('Custom add function message');
          return a + b + c;
        }
      }
      `
    },
    'Curried Function Expression > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = function (a) {
        return function(b) {
          const c = 10;
          scope('Custom add function message');
          return a + b + c;
        }
      }
      `
    },
    'Curried functions > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = function (a) {
        return function(b) {
          return c => {
            const d = 10;
            scope('Custom add function message');
            return a + b + c + d;
          }
        }
      }
      `
    },
    'Return statement should be void when no return found': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const add = function (a, b) {
        scope('Custom add function message');
        a + b;
      }
      `
    },
    'Finds script scope': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const hello = 'hello';
      function divide(a, b) {
        return a / b;
      }
      function add(a, b) {
        scope('Custom add function message');
        return a + b;
      }
      `
    },
    'Export default Function  > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      export default function(state = 0, action) {
        scope('Add one reducer');
        return state + 1;
      }
      `
    },
    'Export default Curried Function  > scope of: custom message, function signature, params, variable, return value and script': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      export default addOne => (state = 0, action) => {
        scope('Add one reducer');
        return addOne(state);
      }
      `
    },
    'Spread arguments': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const sum = (...xs) => {
        scope('Sum args');
        return xs.reduce((acc, x) => acc + x, 0);
      }
      `
    },
    'Multiple scopes': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const sum = (...xs) => {
        scope('Sum args');
        return xs.reduce((acc, x) => acc + x, 0);
      }

      const add = (a, b) => {
        scope('Adding stuff up');
        return a + b;
      }
      `
    },
    'Async functions': {
      snapshot: true,
      code: `
      import scope from './scope.macro';
      const getApi = async (url) => {
        scope('Requesting some data');
        return await fetch(url);
      }
      `
    },
    'module.exports function': {
      skip: true,
      snapshot: true,
      code: `
      import scope from './scope.macro';
      module.exports = (a, b) => {
        scope('Adding stuff up');
        return a + b;
      }
      `
    },
    'Multiple return statements': {
      snapshot: true,
      skip: true,
      code: `
      import scope from './scope.macro';
      const greater = (a, b) => {
        if(a > 1) {
          const c = 'c is for charlie';
          scope('greater than');
          return a;
        }
        const d = 'd is for delta';
        return b;
      }
      `
    },
    'Switch case return': {
      snapshot: true,
      skip: true,
      code: `
      import scope from './scope.macro';
      export default (state = 0, action) => {
        switch(action.type) {
          case 'ADD': {
            scope('Add reducer');
            return add(state, 1);
          }

          default: {
            return 0;
          }
        }
      }
      `
    }
  })
});

function withFilename(tests) {
  return Object.keys(tests).map(key => {
    const t = tests[key];
    const test = { babelOptions: { filename: __filename } };
    if (typeof t === 'string') {
      test.code = t;
    } else {
      Object.assign(test, t);
    }
    return test;
  });
}
