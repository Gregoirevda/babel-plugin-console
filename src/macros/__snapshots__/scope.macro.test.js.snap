// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. macros 1`] = `
"
import scope from './scope.macro'
const c = 10;
scope('should be removed');

      ↓ ↓ ↓ ↓ ↓ ↓

const c = 10;
"
`;

exports[`2. macros 1`] = `
"
import scope from './scope.macro'
function add(a, b) {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

function add(a, b) {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
}
"
`;

exports[`3. macros 1`] = `
"
import scope from './scope.macro'
const add = function (a, b) {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};

      ↓ ↓ ↓ ↓ ↓ ↓

const add = function (a, b) {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};
"
`;

exports[`4. macros 1`] = `
"
import scope from './scope.macro'
const add = (a, b) => {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};

      ↓ ↓ ↓ ↓ ↓ ↓

const add = (a, b) => {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};
"
`;

exports[`5. macros 1`] = `
"
import scope from './scope.macro'
const add = a => b => {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};

      ↓ ↓ ↓ ↓ ↓ ↓

const add = a => b => {
  const c = 10;
  console.log('Custom add function message');
  return a + b + c;
};
"
`;

exports[`6. macros 1`] = `
"
import scope from './scope.macro'
function add(a, b) {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

function add(a, b) {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:0) function add(a, b) {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:13)', 'a:', a);
  console.log('(2:16)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:9)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:0) function add(a, b) {...}');

  return a + b + c;
}
"
`;

exports[`7. macros 1`] = `
"
import scope from './scope.macro';
const add = function(a, b) {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = function (a, b) {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:6) const add = function(a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:21)', 'a:', a);
  console.log('(2:24)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const add = function(a, b) => {...}');

  return a + b + c;
};
"
`;

exports[`8. macros 1`] = `
"
import scope from './scope.macro';
let add = function(a, b) {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

let add = function (a, b) {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:4) let add = function(a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:19)', 'a:', a);
  console.log('(2:22)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:4)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:4) let add = function(a, b) => {...}');

  return a + b + c;
};
"
`;

exports[`9. macros 1`] = `
"
import scope from './scope.macro';
const add = (a, b) => {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = (a, b) => {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:6) const add = (a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:13)', 'a:', a);
  console.log('(2:16)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const add = (a, b) => {...}');

  return a + b + c;
};
"
`;

exports[`10. macros 1`] = `
"
import scope from './scope.macro';
let add = (a, b) => {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

let add = (a, b) => {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:4) let add = (a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:11)', 'a:', a);
  console.log('(2:14)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:4)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:4) let add = (a, b) => {...}');

  return a + b + c;
};
"
`;

exports[`11. macros 1`] = `
"
import scope from './scope.macro';
const add = (a) => (b) => {
  const c = 10;
  scope('Custom add function message');
  return a + b + c;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = a => b => {
  const c = 10;
  console.log('Custom add function message');
  console.groupCollapsed('(2:0) const add = (a) => (b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:13)', 'a:', a);
  console.log('(2:20)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.log('(3:8)', 'c:', c);
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(5:9)', a + b + c);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:0) const add = (a) => (b) => {...}');

  return a + b + c;
};
"
`;

exports[`12. macros 1`] = `
"
import scope from './scope.macro';
function add(a) {
  return (b) => {
    const c = 10;
    scope('Custom add function message');
    return a + b + c;
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

function add(a) {
  return b => {
    const c = 10;
    console.log('Custom add function message');
    console.groupCollapsed('(2:0) const add = (a) => (b) => {...}');
    console.groupCollapsed('Parameters');
    console.log('(2:13)', 'a:', a);
    console.log('(3:10)', 'b:', b);
    console.groupEnd('Parameters');
    console.groupCollapsed('Variables');
    console.log('(4:10)', 'c:', c);
    console.groupEnd('Variables');
    console.groupCollapsed('Return');
    console.log('(6:11)', a + b + c);
    console.groupEnd('Return');
    console.groupCollapsed('Script');
    console.log('(1:7)', 'scope:', scope);
    console.log('(2:9)', 'add:', add);
    console.groupEnd('Script');
    console.groupEnd('(2:0) const add = (a) => (b) => {...}');

    return a + b + c;
  };
}
"
`;

exports[`13. macros 1`] = `
"
import scope from './scope.macro';
const add = function (a) {
  return function(b) {
    const c = 10;
    scope('Custom add function message');
    return a + b + c;
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = function (a) {
  return function (b) {
    const c = 10;
    console.log('Custom add function message');
    console.groupCollapsed('(2:0) const add = (a) => (b) => {...}');
    console.groupCollapsed('Parameters');
    console.log('(2:22)', 'a:', a);
    console.log('(3:18)', 'b:', b);
    console.groupEnd('Parameters');
    console.groupCollapsed('Variables');
    console.log('(4:10)', 'c:', c);
    console.groupEnd('Variables');
    console.groupCollapsed('Return');
    console.log('(6:11)', a + b + c);
    console.groupEnd('Return');
    console.groupCollapsed('Script');
    console.log('(1:7)', 'scope:', scope);
    console.log('(2:6)', 'add:', add);
    console.groupEnd('Script');
    console.groupEnd('(2:0) const add = (a) => (b) => {...}');

    return a + b + c;
  };
};
"
`;

exports[`14. macros 1`] = `
"
import scope from './scope.macro';
const add = function (a) {
  return function(b) {
    return c => {
      const d = 10;
      scope('Custom add function message');
      return a + b + c + d;
    }
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = function (a) {
  return function (b) {
    return c => {
      const d = 10;
      console.log('Custom add function message');
      console.groupCollapsed('(2:0) const add = (a) => (b) => (c) => {...}');
      console.groupCollapsed('Parameters');
      console.log('(2:22)', 'a:', a);
      console.log('(3:18)', 'b:', b);
      console.log('(4:11)', 'c:', c);
      console.groupEnd('Parameters');
      console.groupCollapsed('Variables');
      console.log('(5:12)', 'd:', d);
      console.groupEnd('Variables');
      console.groupCollapsed('Return');
      console.log('(7:13)', a + b + c + d);
      console.groupEnd('Return');
      console.groupCollapsed('Script');
      console.log('(1:7)', 'scope:', scope);
      console.log('(2:6)', 'add:', add);
      console.groupEnd('Script');
      console.groupEnd('(2:0) const add = (a) => (b) => (c) => {...}');

      return a + b + c + d;
    };
  };
};
"
`;

exports[`15. macros 1`] = `
"
import scope from './scope.macro';
const add = function (a, b) {
  scope('Custom add function message');
  a + b;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const add = function (a, b) {
  console.log('Custom add function message');
  console.groupCollapsed('(2:6) const add = function(a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:22)', 'a:', a);
  console.log('(2:25)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('Void');
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const add = function(a, b) => {...}');

  a + b;
};
"
`;

exports[`16. macros 1`] = `
"
import scope from './scope.macro';
const hello = 'hello';
function divide(a, b) {
  return a / b;
}
function add(a, b) {
  scope('Custom add function message');
  return a + b;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const hello = 'hello';
function divide(a, b) {
  return a / b;
}
function add(a, b) {
  console.log('Custom add function message');
  console.groupCollapsed('(6:0) function add(a, b) {...}');
  console.groupCollapsed('Parameters');
  console.log('(6:13)', 'a:', a);
  console.log('(6:16)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(8:9)', a + b);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'hello:', hello);
  console.log('(3:9)', 'divide:', divide);
  console.log('(6:9)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(6:0) function add(a, b) {...}');

  return a + b;
}
"
`;

exports[`17. macros 1`] = `
"
import scope from './scope.macro';
export default function(state = 0, action) {
  scope('Add one reducer');
  return state + 1;
}

      ↓ ↓ ↓ ↓ ↓ ↓

export default function (state = 0, action) {
  console.log('Add one reducer');
  console.groupCollapsed('(2:0) export default function (state, action) {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:24)', 'state:', state);
  console.log('(2:35)', 'action:', action);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(4:9)', state + 1);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.groupEnd('Script');
  console.groupEnd('(2:0) export default function (state, action) {...}');

  return state + 1;
}
"
`;

exports[`18. macros 1`] = `
"
import scope from './scope.macro';
export default addOne => (state = 0, action) => {
  scope('Add one reducer');
  return addOne(state);
}

      ↓ ↓ ↓ ↓ ↓ ↓

export default (addOne => (state = 0, action) => {
  console.log('Add one reducer');
  console.groupCollapsed('(2:0) export default = (addOne) => (state, action) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:15)', 'addOne:', addOne);
  console.log('(2:26)', 'state:', state);
  console.log('(2:37)', 'action:', action);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(4:9)', addOne(state));
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.groupEnd('Script');
  console.groupEnd('(2:0) export default = (addOne) => (state, action) => {...}');

  return addOne(state);
});
"
`;

exports[`19. macros 1`] = `
"
import scope from './scope.macro';
const sum = (...xs) => {
  scope('Sum args');
  return xs.reduce((acc, x) => acc + x, 0);
}

      ↓ ↓ ↓ ↓ ↓ ↓

const sum = (...xs) => {
  console.log('Sum args');
  console.groupCollapsed('(2:6) const sum = (xs) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:16)', 'xs:', xs);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(4:9)', xs.reduce((acc, x) => acc + x, 0));
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'sum:', sum);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const sum = (xs) => {...}');

  return xs.reduce((acc, x) => acc + x, 0);
};
"
`;

exports[`20. macros 1`] = `
"
import scope from './scope.macro';
const sum = (...xs) => {
  scope('Sum args');
  return xs.reduce((acc, x) => acc + x, 0);
}

const add = (a, b) => {
  scope('Adding stuff up');
  return a + b;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const sum = (...xs) => {
  console.log('Sum args');
  console.groupCollapsed('(2:6) const sum = (xs) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:16)', 'xs:', xs);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(4:9)', xs.reduce((acc, x) => acc + x, 0));
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'sum:', sum);
  console.log('(7:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const sum = (xs) => {...}');

  return xs.reduce((acc, x) => acc + x, 0);
};

const add = (a, b) => {
  console.log('Adding stuff up');
  console.groupCollapsed('(7:6) const add = (a, b) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(7:13)', 'a:', a);
  console.log('(7:16)', 'b:', b);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(9:9)', a + b);
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'sum:', sum);
  console.log('(7:6)', 'add:', add);
  console.groupEnd('Script');
  console.groupEnd('(7:6) const add = (a, b) => {...}');

  return a + b;
};
"
`;

exports[`21. macros 1`] = `
"
import scope from './scope.macro';
const getApi = async (url) => {
  scope('Requesting some data');
  return await fetch(url);
}

      ↓ ↓ ↓ ↓ ↓ ↓

const getApi = async url => {
  console.log('Requesting some data');
  console.groupCollapsed('(2:6) const getApi = (url) => {...}');
  console.groupCollapsed('Parameters');
  console.log('(2:22)', 'url:', url);
  console.groupEnd('Parameters');
  console.groupCollapsed('Variables');
  console.groupEnd('Variables');
  console.groupCollapsed('Return');
  console.log('(4:9)', (await fetch(url)));
  console.groupEnd('Return');
  console.groupCollapsed('Script');
  console.log('(1:7)', 'scope:', scope);
  console.log('(2:6)', 'getApi:', getApi);
  console.groupEnd('Script');
  console.groupEnd('(2:6) const getApi = (url) => {...}');

  return await fetch(url);
};
"
`;
